account_invoice_change_currency/models/account_change_currency.py

```diff
@@ -0,0 +1,311 @@
+        if last_currency == self.currency_id and self.get_last_currency_id(True):
+            last_currency = self.get_last_currency_id(True)
+            if last_currency == self.currency_id:
+                last_currency = self.get_last_rate()[0]
+        if (
+            not (self.currency_id and last_currency)
+            or self.currency_id == last_currency
+        ):
+            last_currency = self.company_currency_id
+        today = fields.Date.today()
+        ctx = {"company_id": self.company_id.id, "date": self.invoice_date or today}
+        self.custom_rate = last_currency.with_context(**ctx)._get_conversion_rate(
+            last_currency, self.currency_id, self.company_id, self.invoice_date or today
+        )
+
+    def get_last_currency_id(self, skip_update_currency=False):
+        self.ensure_one()
+        if not self.id:
+            return self.currency_id
+        subtype_id = self.env.ref("account_invoice_change_currency.mt_currency_update")
+        subtype_create_id = self.env.ref("account.mt_invoice_created")
+        query = """
+SELECT mtv.old_value_integer, mtv.new_value_integer, mm.subtype_id
+FROM mail_tracking_value as mtv INNER JOIN mail_message AS mm
+ON mtv.mail_message_id = mm.id INNER JOIN
+ir_model_fields AS field
+ON field.id = mtv.field """
+        params = (self._name, tuple(self.message_ids.ids))
+        if skip_update_currency:
+            query += " AND mm.subtype_id != %s "
+            params = (subtype_id.id, self._name, tuple(self.message_ids.ids))
+        query += """ WHERE field.name = 'currency_id' AND
+field.model = %s AND mtv.mail_message_id IN %s
+ORDER BY mtv.write_date DESC, mtv.id DESC LIMIT 1"""
+        self.env.cr.execute(query, params)
+        res = self.env.cr.dictfetchone()
+        value = False
+        if res:
+            value = res["old_value_integer"]
+            if res["subtype_id"] == subtype_create_id.id:
+                value = res["new_value_integer"]
+        return self.currency_id.browse(value)
+
+    def get_last_rate(self):
+        self.ensure_one()
+        subtype_create_id = self.env.ref("account.mt_invoice_created")
+        ir_model_fields = self.env["ir.model.fields"]
+        ir_model_fields_ids = (
+            ir_model_fields.sudo()
+            .search(
+                [
+                    ("name", "in", ["rate", "currency_id"]),
+                    ("model", "in", [self._name, "res.currency"]),
+                ]
+            )
+            .ids
+        )
+        last_values = (
+            self.env["mail.tracking.value"]
+            .sudo()
+            .search(
+                [
+                    ("mail_message_id", "in", self.message_ids.ids),
+                    ("field", "in", ir_model_fields_ids),
+                ],
+                limit=2,
+                order="write_date desc, id desc",
+            )
+        )
+        if len(last_values) == 2:
+            currency_value, rate_value = sorted(
```

Comment:
This seems too complicated

Reply:
I made a change with this, you could review it